Main/Workshop3/DeckBase.h

```diff
@@ -0,0 +1,40 @@
+#pragma once
+
+#include <vector>
+#include <memory>
+#include <random>
+#include <algorithm>
+
+// Template class for generic deck
+template <typename T>
+class DeckBase {
+public:
+    DeckBase();
+    void addElement(std::unique_ptr<T> element);
+    std::unique_ptr<T> generateElement();
+    size_t size() const { return elements.size(); }
+
+private:
+    std::vector<std::unique_ptr<T>> elements;
+};
+
+template <typename T>
+DeckBase<T>::DeckBase() = default;
+
+template <typename T>
+void DeckBase<T>::addElement(std::unique_ptr<T> element) {
+    elements.push_back(std::move(element));
+}
+
+template <typename T>
+std::unique_ptr<T> DeckBase<T>::generateElement() {
+    if (elements.empty()) return nullptr;
```

Comment:
And when the deck is empty you just return nullptr?
More flexible approach would be to regenerate the deck once again. Aka similarily it's done in real game. When all cards where played you just reshuffle and play them once more